jetty-io/src/main/java/org/eclipse/jetty/io/MemoryPool.java

```diff
@@ -0,0 +1,53 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import org.eclipse.jetty.util.component.Container;
+
+/**
+ * A Pool of objects representing memory that can be acquired based on size and direction. The held instances may be the memory
+ * component itself (e.g.: {@link java.nio.ByteBuffer}) or an abstraction providing access to the memory component.
+ * @param <T> The memory buffer type.
+ */
+public interface MemoryPool<T>
+{
+    /**
+     * Acquire a memory buffer from the pool.
+     * @param size The size of the buffer. The returned buffer will have at least this capacity.
+     * @param direct true if a direct memory buffer is needed, false otherwise.
+     * @return a memory buffer.
+     */
+    T acquire(int size, boolean direct);
+
+    /**
+     * Release a previously acquired memory buffer to the pool.
+     * @param buffer the memory buffer to release.
+     */
+    void release(T buffer);
+
+    /**
+     * Find a {@link MemoryPool} of {@link RetainableByteBuffer} implementation in the given container, or wrap the given
+     * {@link ByteBufferPool} with an adapter.
+     * @param container the container to search for an existing memory pool.
+     * @param byteBufferPool the {@link ByteBufferPool} to wrap if no memory pool was found in the container.
+     * @return the memory pool found or the wrapped one.
+     */
+    static MemoryPool<RetainableByteBuffer> findOrAdapt(Container container, ByteBufferPool byteBufferPool)
```

Comment:
I like the method, but it should be on `RetainableByteBufferPool`. I think that would allow AdaptorMemoryPool to be entirely private.

Also, perhaps there should be an option or variant to create a real RetainableByteBufferPool and add it to the container.   In websocket we name such methods `ensureRetainableByteBufferPool`.

Reply:
Agreed that this method should move to `RetainableByteBufferPool` to allow `AdapterMemoryPool` to be private.

I'm also going to have a look at what `ensureRetainableByteBufferPool()` would look like.

Reply:
`ensureRetainableByteBufferPool` doesn't make much sense IMHO as there only are two places where `RetainableByteBufferPool` is instantiated and they both use different constructors and do not lookup and register in containers in the same way.